### PR TITLE
Display message if a direct-linked annotation is not available

### PR DESCRIPTION
**Pull request by robertknight**
This is based on **#3108** and will need to be rebased and merged afterwards.

This adds a message in the sidebar which is displayed if a direct-linked annotation is selected but that annotation is not available, after fetching all annotations for the current page.

### CODE REVIEW

**Comment by robertknight** [comment]
<a href="https:&#x2F;&#x2F;trello.com&#x2F;c&#x2F;0dlSUhwO&#x2F;256-allow-a-url-fragment-to-trigger-the-client-to-focus-a-specific-annotation-on-load"> Allow a URL fragment to trigger the client to focus a specific annotation on load.</a>

File: h/static/styles/thread.scss [hunk 1]
```diff
@@ -15,6 +15,26 @@ $thread-padding: $annotation-card-left-padding;
   }
 }
 
+.annotation-unavailable-message {
+  display: flex;
+  flex-direction: column;
+  border: 1px solid $gray-lighter;
+  padding-top: 30px;
```
Comment (robertknight) [inline h/static/styles/thread.scss]: TODO: The multiples of 10px here are chosen to align with the 10px grid which the new design is based off. Now would be a good time to introduce that as a variable in `variables.scss`

File: h/static/scripts/search-client.js [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+ * SearchClient handles paging through results, canceling search etc.
+ *
+ * @param {Object} resource - ngResource class instance for the /search API
+ * @param {Object} opts - Search options
+ * @constructor
+ */
+function SearchClient(resource, opts) {
+  opts = opts || {};
+
+  var DEFAULT_CHUNK_SIZE = 200;
```
Comment (nickstenning) [inline h/static/scripts/search-client.js]: This has increased from 100 to 200 -- that is possibly ok, but it might be better to leave it as is for now because of [the rather ugly way replies are loaded on the server](https://github.com/hypothesis/h/blob/3e8d22814a4e3ce8e6e743d5ca1ae59820c5dbe0/h/api/search/core.py#L58).

File: h/static/scripts/groups.js [hunk 3]
```diff
@@ -72,12 +72,15 @@ function groups(localStorage, session, settings, $rootScope, $http) {
...
-     focusedGroup = g;
-     localStorage.setItem(STORAGE_KEY, g.id);
-     $rootScope.$broadcast(events.GROUP_FOCUSED, g.id);
-   }
+    var prevFocused = focused();
+    var g = get(id);
+    if (g) {
+      focusedGroup = g;
+      localStorage.setItem(STORAGE_KEY, g.id);
+      if (prevFocused.id !== g.id) {
```
Comment (nickstenning) [inline h/static/scripts/groups.js]: This change seems a bit out of place in this commit. It certainly looks harmless, but I don't see how it has anything to do with pagination of search results, so either an explanation in the commit message, or splitting it into its own commit wouldn't go amiss.

**Comment by nickstenning** [comment]
Ah, just realised I'm reviewing these in the wrong order. Switching to #3108.

File: h/static/scripts/widget-controller.js [hunk 4]
```diff
@@ -193,6 +203,12 @@ module.exports = function WidgetController(
     return annotation.$$tag in $scope.focusedAnnotations;
   };
 
+  $scope.selectedAnnotationUnavailable = function () {
+    return searchClients.length === 0 &&
```
Comment (nickstenning) [inline h/static/scripts/widget-controller.js]: Related to a comment I made on #3108, relying on `searchClients.length` to determine whether requests are in-flight to fetch annotations is pretty opaque. At the very least add a comment, or better add a helper function elsewhere in this file to hide the details of `searchClients` from other parts of the code.

File: h/templates/client/viewer.html [hunk 5]
```diff
@@ -21,6 +21,13 @@
       on-change-sort-by="sort.name = sortBy">
     </sort-dropdown>
   </li>
+  <li class="annotation-unavailable-message"
+      ng-if="selectedAnnotationUnavailable()">
+    <span class="h-icon-lock annotation-unavailable-message__icon"></span>
```
Comment (nickstenning) [inline h/templates/client/viewer.html]: This icon isn't the one on the Trello card. Is that agreed on?

File: h/static/scripts/groups.js (hunk 3 above)
Comment (robertknight) [inline h/static/scripts/groups.js]: I'll add an explanation here - the gist is that focusing a group triggers a reload of annotations in `widget-controller.js`, but we don't want to do that unnecessarily if `groups.focus()` is called with a group that is the same as the current group.

Comment (robertknight) [inline h/static/scripts/groups.js]: Update - I split this into a separate commit in #3108 and added an explanation plus tests.

File: h/templates/client/viewer.html (hunk 5 above)
Comment (robertknight) [inline h/templates/client/viewer.html]: No. It needs to be updated to use the SVG that Conor provided.

**Comment by nickstenning** [comment]
Ok, then checklist to get this merged:

- [ ] uses correct icon for the padlock

**Comment by nickstenning** [comment]
LGTM.
